debezium-core/src/main/java/io/debezium/connector/common/BaseSourceTask.java

```diff
@@ -107,8 +139,20 @@ public void commitRecord(SourceRecord record) throws InterruptedException {
 
     @Override
     public void commit() throws InterruptedException {
-        if (coordinator != null && lastOffset != null) {
-            coordinator.commitOffset(lastOffset);
+        boolean locked = stateLock.tryLock();
+
+        if (locked) {
+            try {
+                if (coordinator != null && lastOffset != null) {
+                    coordinator.commitOffset(lastOffset);
+                }
+            }
+            finally {
+                stateLock.unlock();
+            }
+        }
+        else {
+            LOGGER.warn("Couldn't commit processed log positions with the source database due to a concurrent connector shutdown or restart");
```

Comment:
Does this mean that during stopping a Connect Worker without stopping the connector first will mean that all queued records will not get flushed and will be retried when the connector restarts?

Reply:
Nevermind, got it. You reload the offsets from the in-memory state at restart to avoid repeated flushes.